### PR TITLE
Feature/vault

### PR DESCRIPTION
**Pull request by Matt-Guest**
### What

Created new Vault client library in go-ns. Includes some documentation and examples using and generating tokens.

### How to review

Read the documentation, review the code, run the code.

### Who can review

anyone

### CODE REVIEW

File: vault/client.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	config := vaultapi.Config{Address: vaultAddress, MaxRetries: retries}
+	client, err := vaultapi.NewClient(&config)
+	if err != nil {
+		return nil, err
+	}
+	client.SetToken(token)
+	return &VaultClient{token, client}, nil
+}
+
+// CreateTLSVaultClient by providing a auth token, vault address and the maxium number of retries for a request
```
Comment (gedge) [inline vault/client.go]: this name irks me - does `CreateVaultClientTLS` sound better to you?

File: vault/client.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	vaultapi "github.com/hashicorp/vault/api"
+)
+
+// VaultClient Used to read and write secrets from vault
+type VaultClient struct {
+	token  string
+	client *vaultapi.Client
+}
+
+// CreateVaultClient by providing a auth token, vault address and the maxium number of retries for a request
```
Comment (gedge) [inline vault/client.go]: typos `a auth token` and `maxium`?

File: vault/client.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	"errors"
+
+	vaultapi "github.com/hashicorp/vault/api"
+)
+
+// VaultClient Used to read and write secrets from vault
+type VaultClient struct {
+	token  string
```
Comment (gedge) [inline vault/client.go]: is `token` ever used (e.g. by a getter)?

File: vault/client.go (hunk 1 above)
Comment (gedge) [reply vault/client.go]: same typos as above

File: vault/client.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	client.SetToken(token)
+	return &VaultClient{token, client}, nil
+}
+
+// Read a secret from vault. If the token does not have the correct policy this shall return an error or if the
+// vault server is not reachable. This method shall return all the information store about the secret
```
Comment (gedge) [inline vault/client.go]: did you mean
```
// Read reads a secret from vault. If the token does not have the correct policy this returns an error;
// if the vault server is not reachable, return all the information stored about the secret.
```

File: vault/client.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	val, ok := data[key]
+	if !ok {
+		return "", errors.New("key not found")
+	}
+	return val.(string), nil
+}
+
+// Write a secret to vault. If the token does not have the correct policy this shall return an error or if the
+// vault server is not reachable. If nil is returned the operation was successful
```
Comment (gedge) [inline vault/client.go]:
```
// Write writes a secret to vault. Returns an error if the token does not have the correct policy or the
// vault server is not reachable. Returns nil when the operation was successful.
```

File: vault/example/Makefile [hunk 6]
```diff
@@ -0,0 +1,22 @@
+
+VAULT_ADDR='http://127.0.0.1:8200'
```
Comment (gedge) [inline vault/example/Makefile]: make this conditional (defaulted to) with `VAULT_ADDR?='http://127.0.0.1:8200'`

File: vault/example/example.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/ONSdigital/go-ns/vault"
+)
+
+func main() {
+
+	log.Namespace = "vault-example"
+	log.Debug("Creating vault client", log.Data{})
+	devAddress := "http://127.0.0.1:8200"
```
Comment (gedge) [inline vault/example/example.go]: this should be `vaultAddress := os.Getenv("VAULT_ADDR")`

File: vault/client.go (hunk 1 above)
Comment (Matt-Guest) [reply vault/client.go]: I renamed the function and fixed the typo

File: vault/client.go (hunk 2 above)
Comment (Matt-Guest) [reply vault/client.go]: done

File: vault/client.go (hunk 3 above)
Comment (Matt-Guest) [reply vault/client.go]: No its never used. I removed it 👍

File: vault/client.go (hunk 4 above)
Comment (Matt-Guest) [reply vault/client.go]: Thanks

File: vault/client.go (hunk 5 above)
Comment (Matt-Guest) [reply vault/client.go]: 👍

File: vault/example/Makefile (hunk 6 above)
Comment (Matt-Guest) [reply vault/example/Makefile]: done

File: vault/example/example.go (hunk 7 above)
Comment (Matt-Guest) [reply vault/example/example.go]: done

**Comment by Matt-Guest** [comment]
Is been vendored already? Is there anything missing you can see?

**Pull request merged**